keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,776 @@
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### New Resource
+```
+ClusterCIDRConfig {
+    # Nodes may only have 1 range from each family.
+    # TODO: maybe this is redundant and can just be inferred by the controller.
+    IPFamily        FamilyType
+    
+    # An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64")
+    # +required
+    IPCIDRBlock     string
+    
+    # This defines which nodes the config is applicable to. An empty selector
+    # can be applied to any node.
+    # +optional
+    NodeSelector    v1.LabelSelector
+
+    # Netmask size (e.g. 25 -> "/25") to allocate to a node.
+    # Users would have to ensure that the kubelet doesn't try to schedule 
+    # more pods than are supported by the node's netmask (i.e. the kubelet's
+    # --max-pods flag)
+    # +required
+    PerNodeMaskSize int
+}
+
+var (
+    IPV4 FamilyType = "IPV4"
+    IPV6 FamilyType = "IPV6"
+)
+```
+
+#### Expected Behavior
+
+- Each node will be assigned up to one range from each `FamilyType`.
+    - **TODO:** what is the tie-break if multiple ranges match
+- An empty `NodeSelector` functions as a default that applies to all nodes. This should be the fall-back and not take precedence if any other range matches.
+- `IPFamily`, `IPCIDRBlock`, and `PerNodeMaskSize` are immutable after creation.
+- Reject any changes that would change existing node allocations:
+    - Can not delete a `CIDRRangeConfig` that is in use by a node
+    - Can not modify a `NodeSelector` to deselect a node that is using that range.
+
+#### Example Config
+```
+[
+{
+    IPFamily:        IPV4,
+    # For existing clusters this is the same as ClusterCIDR
+    IPCIDRBlock:     "10.0.0.0/8",
+    # Default for nodes not matching any other rule
+    NodeSelector:    nil,
+    # For existing API this is the same as NodeCIDRMaskSize
+    PerNodeMaskSize: 24,
+},
+{
+    IPFamily:        IPV4,
+    IPCIDRBlock:     "172.16.0.0/14",
+    # Another range, also allocate-able to any node
+    NodeSelector:    nil,
+    PerNodeMaskSize: 24,
+},
+{
+    IPFamily:        IPV4,
+    IPCIDRBlock:     "10.0.0.0/8",
+    NodeSelector:    { key: "np" op: "IN" value:["np1"] },
+    PerNodeMaskSize: 26,
+},
+{
+    IPFamily:        IPV4,
+    IPCIDRBlock:     "192.168.0.0/16",
+    NodeSelector:    { key: "np" op: "IN" value:["np2"] },
+    PerNodeMaskSize: 26,
+},
+{
+    IPFamily:        IPV4,
+    IPCIDRBlock:     "5.2.0.0/16",
+    NodeSelector:    { "np": "np3" },
+    PerNodeMaskSize: 20,
+},
+{
+    IPFamily:        IPV6,
+    IPCIDRBlock:     "fd12:3456:789a:1::/64",
+    NodeSelector:    { "np": "np3" },
+    PerNodeMaskSize: 112,
+},
+...
+]
+```
+
+Given the above config, a valid potential configuration might be:
+
+```
+{“np”: “np1”} --> "10.0.0.0/26"
+{“np”: “np2”} --> "192.16.0.0/26"
+{“np”: “np3”} --> "5.2.0.0/20", "fd12:3456:789a:1::/112"
+{“np”: “np4”} --> "172.16.0.0/24"
```

Comment:
hmm, will this break the premise that all pods should be able to reach all pods in the cluster?
should we allow this mix or "enforce" all nodes to have the same families?
I prefer the later

Reply:
Agree, I think.  It's a little daunting to expose so much of the inner workings, I admit.  I think we would strongly urge people to create pairs of configs in dual-stack mode.  IIRC there's not a "primary family" notion for pod CIDRs like there is with Services.